federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+		cachedService.serviceStatusMap[clusterName] = service.Status.LoadBalancer
+		needUpdate = true
+
+		// cache is not always reliable(cache will be cleaned when service controller restart)
+		// two cases will run into this branch:
+		// 1. new service loadbalancer info received -> no info in cache, and no in federation service
+		// 2. service controller being restarted -> no info in cache, but it is in federation service
+
+		// check if the lb info is already in federation service
+		CheckLB:
+			for _, fed := range cachedFedServiceStatus.Ingress {
+				for _, new := range service.Status.LoadBalancer.Ingress{
+					if (new.IP == fed.IP && new.Hostname == fed.Hostname) {
+						needUpdate = false
+						break CheckLB
+					}
+				}
+			}
+	}
+
+	if needUpdate {
+		// new status = cached federation status - cached status for the cluster + new status from the cluster
+		for i, fed := range cachedFedServiceStatus.Ingress {
+			for _, new := range cachedStatus.Ingress{
+				// remove if same ingress record found
+				if (new.IP == fed.IP && new.Hostname == fed.Hostname) {
+					cachedFedServiceStatus.Ingress = append(cachedFedServiceStatus.Ingress[:i], cachedFedServiceStatus.Ingress[i+1:]...)
+					glog.V(4).Infof("remove old ingress info %s, %s for service %s/%s", new.Hostname, new.IP, service.Namespace, service.Name)
+				}
+			}
+		}
+		cachedService.lastState.Status.LoadBalancer.Ingress = append(cachedFedServiceStatus.Ingress, service.Status.LoadBalancer.Ingress...)
+		glog.V(4).Infof("add new ingress info %+v for service %s/%s", service.Status.LoadBalancer, service.Namespace, service.Name)
+	} else {
+		glog.V(4).Infof("same ingress info found for %s/%s, cluster %s", service.Namespace, service.Name, clusterName)
+	}
+	return needUpdate
+}
+
+
+func (cc *clusterClientCache) persistFedServiceUpdate(service *v1.Service, fedClient federation_release_1_3.Interface) (error, bool){
+	glog.V(5).Infof("Persist federation service status %+v", service)
+	for i := 0; i < clientRetryCount; i++ {
+		_, err := fedClient.Core().Services(service.Namespace).UpdateStatus(service)
+		if err == nil {
+			glog.V(2).Infof("Successfully update service %s/%s to federation apiserver", service.Namespace, service.Name)
+			return nil, !retryable
+		}
+		if errors.IsNotFound(err) {
+			glog.Infof("Not persisting update to service '%s/%s' that no longer exists: %v",
+				service.Namespace, service.Name, err)
+			return nil, !retryable
+		}
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			return nil, !retryable
+		}
+		glog.Warningf("Failed to persist updated LoadBalancerStatus to service '%s/%s' after creating its DNS records: %v",
+			service.Namespace, service.Name, err)
+		time.Sleep(clientRetryInterval)
+	}
+	return nil, !retryable
```

Comment:
I might be missing something here, but it seems like once we fail after clientRetyCount attempts, we never retry again, and the state of the federated service is potentially out of date forever?